### PR TITLE
Features/identity contract 178007199

### PR DESCRIPTION
**Pull request by aneyzberg**
Problem
=======
We want to be able to create clone proxy contracts and beacon proxy contracts from the SDK
[link to Pivotal Tracker #178007199](https://www.pivotaltracker.com/story/show/178007199)

Solution
========
Add functionality to create clone proxy contracts and beacon proxy contracts to the SDK

with @rosalinekarr 

Change summary:
---------------
* Add functions to `identity.ts` to create both clone and beacon proxy contracts
* Add function to check authorization on proxy contract
* Add `identity.test.ts` spec
* Refactor Announcer to no longer use `getContract`
* retrieve contracts using the typechain factories rather than the `getContract` method
* Update SDK to latest version of contracts repo - `0.0.0-9c117e`

### CODE REVIEW

File: src/contracts/identity.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+export const isAuthorizedTo = async (
+  address: EthereumAddress,
+  contractAddress: EthereumAddress,
+  permission: Permission,
+  blockNumber: number,
+  opts?: Config
+): Promise<boolean> => {
+  const { provider, signer } = await getConfig(opts);
+
+  if (!signer) throw MissingSigner;
```
Comment (wilwade) [inline src/contracts/identity.ts]: This is just a call, it should not need the `signer`.

File: src/contracts/identity.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+ */
+export const createBeaconProxyWithOwner = async (
+  owner: EthereumAddress,
+  beacon: EthereumAddress
+): Promise<ContractTransaction> => {
+  const contract = await getBeaconFactoryContract();
+  return contract.createBeaconProxyWithOwner(beacon, owner);
+};
+
+const getIdentityContractAddress = async (): Promise<EthereumAddress> => {
```
Comment (wilwade) [inline src/contracts/identity.ts]: Naming?
```suggestion
const getIdentityLogicContractAddress = async (): Promise<EthereumAddress> => {
```

File: src/config/config.ts [hunk 3]
```diff
@@ -11,6 +11,12 @@ export interface Config {
   queue: QueueInterface;
   store?: StorageInterface;
   contracts: {
+    announcer?: HexString;
+    beacon?: HexString;
+    beaconFactory?: HexString;
+    identity?: HexString;
```
Comment (wilwade) [inline src/config/config.ts:17]: Name suggestion
```suggestion
    identityLogic?: HexString;
```

File: src/contracts/contract.ts [hunk 4]
```diff
@@ -3,6 +3,7 @@ import { HexString } from "../types/Strings";
 import { keccak256 } from "js-sha3";
 import { ethers } from "ethers";
 
+export const GAS_LIMIT_BUFFER = 1000;
```
Comment (wilwade) [inline src/contracts/contract.ts]: Do we need this anymore?

**Review by wilwade — CHANGES_REQUESTED** [review]

File: package-lock.json [hunk 5]
```diff
@@ -626,47 +623,49 @@
...
-      "integrity": "sha512-sGkPx+VjMtmA6MX27oA4FBFELFCZZ4S4XqeGOXCv68tT+jb3vk/RyaKWP0PTKyWtmLSM0b+adUTEvbs1PEaH2w==",
-      "dev": true
+      "dev": true,
+      "license": "MIT"
     },
     "node_modules/@ethersproject/abi": {
-      "version": "5.3.0",
-      "resolved": "https://registry.npmjs.org/@ethersproject/abi/-/abi-5.3.0.tgz",
-      "integrity": "sha512-NaT4UacjOwca8qCG/gv8k+DgTcWu49xlrvdhr/p8PTFnoS8e3aMWqjI3znFME5Txa/QWXDrg2/heufIUue9rtw==",
+      "version": "5.1.2",
```
[the remark this answers is not in the document]
Comment (aneyzberg) [reply package-lock.json:668]: my mistake in rebase.

File: src/contracts/identity.ts (hunk 1 above)
Comment (aneyzberg) [reply src/contracts/identity.ts]: yes. my mistake. Meant to remove.

File: src/contracts/identity.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  if (!signer) throw MissingSigner;
+  if (!provider) throw MissingProvider;
+
+  const address = beaconFactory || (await getContractAddress(provider, BEACON_CONTRACT));
+  if (!address) throw MissingContract;
+
+  return BeaconFactory__factory.connect(address, signer);
+};
+
+export const isAuthorizedTo = async (
```
[the remark this answers is not in the document]
Comment (aneyzberg) [reply src/contracts/identity.ts:122]: this is resolved if you look at the file on the branch. Github seems to be having an issue picking this up.

**Review by wilwade — APPROVED** [review]
🚂

**Pull request merged**